### DISCUSSION TITLE
Changelog propagation for regex manager with conan datasource

### DISCUSSION THREAD

**Discussion by MartijnLeijssen-TomTom** [renovatebot/renovate#22625] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Latest (renovatebot/github-action@v38.1.2)

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

`conan` datasource to maintain our conan packages. Renovate is capable of forwarding the changelog based on github releases which is exactly what we want, but  the conan datasource doesn't support this feature. In our company we are utilizing artifactory to upload the conan packages and on artifactory the "sourceUrl" of the package is set as an property on the artifact (on the recipe to be more precise). So we basically have all the information available to make the changelog work, but I don't see a good way of implementing this is a way it could be upstreamed, as it will combine artifactory and conan rest api calls.

### What you tried so far.

I modified the conan datasource to get the latest version of each package, get the latest uploaded revision of the version and read the corresponding url property from artifactory (`conan.package.url`). The code works properly, but I don't think it is a good idea the couple the artifactory api with the conan rest api in the conansource. I could create a "artifactory conan" datasource which would be mostly the same as a conan manager, with additional logic to fetch the changelog url, but not sure if this is acceptable.

- Is there a way to add custom datasources ourself?
- Is there a suggestion how to make desired changed as a feature flag/configuration flag? 

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
Is Artifactory behaving as a "conan compatible" datasource? And if so then when Renovate queries it to look up a Conan package, is that source URL included in the response - or does it require a separate API query?

**Reply by MartijnLeijssen-TomTom** [reply]
> Is Artifactory behaving as a "conan compatible" datasource? 

Yes

>  is that source URL included in the response - or does it require a separate API query?

Separate queries, that is the reason I have some doubts making this change. 
It actually requires 2 API calls per package:
Get the latest package revision (not to be confused with latest version, which can be determined already with the information that the datasource has). After the revision is known, the sourceUrl can be queried from the corresponding recipe
1:  get revision: v2/conans/${conanPackage.conanName}/${latestVersion}/${conanPackage.userAndChannel}/latest
2:  get url: api/storage/${repo}/${user}/${conanPackage.conanName}/${latestVersion}/${channel}/${packageRev}/export/conanfile.py?properties=conan.package.url

[EDIT]
You can see my [experimental code ](https://github.com/MartijnLeijssen-TomTom/renovate/pull/new/feat/changelog_conan) (this works for the use case that I want to solve)

**Reply by viceice** [reply]
we could maybe add a condition. so when we fetch the new versions, we can check for the artifactory header and do the additional API request to get the source url.

we do something similar inside docker datasource.

**Comment by MartijnLeijssen-TomTom** [comment]
Thanks! I followed that approached and raised an [PR](https://github.com/renovatebot/renovate/pull/22656) for this functionality, let me know what you think!
